clang/lib/Analysis/FlowSensitive/Models/ReverseNullModel.cpp

```diff
@@ -0,0 +1,723 @@
+        // FIXME: Hack
+        if (!EnvVal->getProperty(kIsNull) && !EnvVal->getProperty(kIsNonnull)) {
+            initializeRootValue(*EnvVal, Env);
+        }
+
+        return EnvVal;
+    }
+
+    auto *RootValue = Env.createValue(Var.getType());
+
+    if (!RootValue) 
+        return nullptr;
+
+    initializeRootValue(*RootValue, Env);
+
+    setLValue(Var, *RootValue, Env);
+
+    return RootValue;
+}
+
+void matchDereferenceExpr(const Stmt *stmt,
+                          const MatchFinder::MatchResult &Result,
+                          Environment &Env) {
+    const auto *Var = Result.Nodes.getNodeAs<Expr>(kVar);
+    assert(Var != nullptr);
+
+    // FIXME: Optimize
+    auto *RootValue = getValue(*Var, Env);
+    if (!RootValue) {
+       return;
+    }
+
+    Env.addToFlowCondition(Env.arena().makeNot(getVal(kIsNull, *RootValue).formula()));
+}
+
+void matchCastExpr(const CastExpr *cond, const MatchFinder::MatchResult &Result,
+                   ReverseNullModel::TransferArgs &Data) {
+    auto IsNonnullBranch = Data.Branch;
+    auto &Env = Data.Env;
+
+    const auto *Var = Result.Nodes.getNodeAs<Expr>(kVar);
+    assert(Var != nullptr);
+
+    auto *RootValue = getValue(*Var, Env);
+    if (!RootValue) {
+       return;
+    }
+
+    auto *NewRootValue = Env.createValue(Var->getType());
+    if (!NewRootValue) 
+        return;
+
+    setLValue(*Var, *NewRootValue, Env);
+
+    Arena &A = Env.arena();
+
+    if (IsNonnullBranch) {
+        BoolValue &IsNonnull = getVal(kIsNonnull, *RootValue);
+        BoolValue &IsNull = getVal(kIsNull, *RootValue);
+
+        Env.addToFlowCondition(A.makeNot(IsNull.formula()));
+        NewRootValue->setProperty(kIsNull, Env.getBoolLiteralValue(false));
+
+        NewRootValue->setProperty(kIsNonnull, IsNonnull);
+    } else {
+        BoolValue &IsNonnull = getVal(kIsNonnull, *RootValue);
+        BoolValue &IsNull = getVal(kIsNull, *RootValue);
+
+        NewRootValue->setProperty(kIsNull, IsNull);
+
+        Env.addToFlowCondition(A.makeNot(IsNonnull.formula()));
+        NewRootValue->setProperty(kIsNonnull, Env.getBoolLiteralValue(false));
+    }
+}
+
+void matchNullptrExpr(const Expr *expr, const MatchFinder::MatchResult &Result, Environment &Env) {
+    const auto *PrVar = Result.Nodes.getNodeAs<Expr>(kVar);
+    assert(PrVar != nullptr);
+
+    if (Env.getValue(*PrVar))
+      return;
+
+    // FIXME: Nullptr does not create a value
+    auto *RootValue = Env.createValue(PrVar->getType());
+    if (!RootValue) {
+       return;
+    }
+
+    RootValue->setProperty(kIsNull, Env.getBoolLiteralValue(true));
+    RootValue->setProperty(kIsNonnull, Env.getBoolLiteralValue(false));
+    Env.setValue(*PrVar, *RootValue);
+}
+
+void matchAddressofExpr(const Expr *expr, const MatchFinder::MatchResult &Result, Environment &Env) {
+    const auto *PrVar = Result.Nodes.getNodeAs<Expr>(kVar);
+    assert(PrVar != nullptr);
+
+    auto *RootValue = Env.createValue(PrVar->getType());
```

Comment:
Is the type of the... pointer...(?) a key here somehow? Why is this distinction relevant, when the values are set attached to the expression node? What happens if we have two pointers of the same type?